README
Calendar-Saka

The README  is  used  to  introduce the module and provide instructions on how to
install the module, any machine dependencies it may have (for example C compilers
and installed libraries) and any other information that should be provided before
the module is installed.

A  README  file  is required for CPAN modules since CPAN extracts the README file
from a module distribution so that people browsing the archive can use it to  get
an  idea  of  the  module's  uses.  It  is usually a good idea to provide version
information here so that people can decide whether fixes for the module are worth
downloading.

INSTALLATION

To install this module, run the following commands:

	perl Makefile.PL
	make
	make test
	make install

SUPPORT AND DOCUMENTATION

After  installing, you  can  find documentation for this module with  the perldoc
command.

    perldoc Calendar::Saka

You can also look for information at:

    RT, CPAN's request tracker
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=Calendar-Saka

    AnnoCPAN, Annotated CPAN documentation
        http://annocpan.org/dist/Calendar-Saka

    CPAN Ratings
        http://cpanratings.perl.org/d/Calendar-Saka

    Search CPAN
        http://search.cpan.org/dist/Calendar-Saka/

LICENSE AND COPYRIGHT

Copyright (C) 2011 - 2015 Mohammad S Anwar

This program is free software; you can redistribute it and/or modify it under the
terms of the the Artistic License (2.0).You may obtain a copy of the full license
at:

L<http://www.perlfoundation.org/artistic_license_2_0>

Any  use,  modification, and distribution of the Standard or Modified Versions is
governed by this Artistic License.By using, modifying / distributing the Package,
you accept this license. Do not use, modify, or distribute the Package, if you do
not accept this license.

If your Modified Version has been derived from a Modified Version made by someone
other than you, you're nevertheless required to ensure that your Modified Version
complies with the requirements of this license.

This  license  does  not  grant you the right to use any trademark, service mark,
tradename, or logo of the Copyright Holder.

This license includes the non-exclusive, worldwide, free-of-charge patent license
to  make,  have made, use, offer to sell, sell, import and otherwise transfer the
Package with respect to any patent claims licensable by the Copyright Holder that
are  necessarily  infringed  by the  Package.  If you institute patent litigation
(including  a  cross-claim or  counterclaim)  against any party alleging that the
Package constitutes direct or contributory patent infringement,then this Artistic
License to you shall terminate on the date that such litigation is filed.

Disclaimer  of  Warranty:  THE  PACKAGE  IS  PROVIDED BY THE COPYRIGHT HOLDER AND
CONTRIBUTORS  "AS IS'  AND WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES. THE IMPLIED
WARRANTIES    OF     MERCHANTABILITY,   FITNESS   FOR   A PARTICULAR  PURPOSE, OR
NON-INFRINGEMENT ARE DISCLAIMED TO THE EXTENT PERMITTED BY YOUR LOCAL LAW. UNLESS
REQUIRED BY LAW, NO COPYRIGHT HOLDER OR CONTRIBUTOR WILL BE LIABLE FOR ANY DIRECT,
INDIRECT, INCIDENTAL, OR CONSEQUENTIAL DAMAGES ARISING IN ANY WAY OUT OF THE USE
OF THE PACKAGE,EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

Changes
Revision history for Calendar::Saka

1.33  Tue Jan 10 10:50:00 2017
      - Added dependency on Calendar::Plugin::Renderer v0.12.
      - Removed method validate_params(), no longer required.

1.32  Tue Jan 03 07:55:00 2017
      - Added sample code for method as_text() in the SYNOPSIS section.
      - Updated pod document for method as_text().

1.31  Sat Dec 10 12:35:00 2016
      - Added test scripts (t/meta-json.t and t/meta-yml.t).

1.30  Mon Nov 14 09:20:00 2016
      - Upgraded to use Date::Saka::Simple v0.17.

1.29  Thu Oct 20 09:55:00 2016
      - Added dependency on Calendar::Plugin::Renderer v0.11.

1.28  Mon Sep 26 11:00:00 2016
      - Added key 'provides' to the Makefile.PL script.

1.27  Fri Aug 12 12:30:00 2016
      - Upgraded to use Date::Saka::Simple v0.15.

1.26  Sat Jul 09 08:00:00 2016
      - Added method as_text().
      - Updated method as_svg() with regard to start index.
      - Added dependency on Calendar::Plugin::Renderer v0.07.
      - Added test for method as_svg().

1.25  Tue May 24 10:50:00 2016
      - Upgraded to use Date::Saka::Simple v0.12.
      - Updated copyright year information.
      - Updated method as_svg() to handle month name.
      - Refactored method current() and method from_gregorian().
      - Updated pod document with regard to the Calendar::Plugin::Renderer v0.06.

1.24  Mon Jan 11 09:45:00 2016
      - Added method as_svg() using the plugin Calendar::Plugin::Renderer.

1.23  Sat Aug 29 14:55:00 2015
      - Upgraded to use Date::Saka::Simple v0.09.

1.22  Tue Aug 11 11:45:00 2015
      - Upgraded to use Date::Saka::Simple v0.08.
      - Making use of the method get_calendar() provided by Date::Saka::Simple.

1.21  Fri Jul 10 10:05:00 2015
      - Upgraded to use Date::Saka::Simple v0.07.

1.20  Tue Jun 30 11:00:00 2015
      - Upgraded the dependency on Date::Saka::Simple v0.06.
      - Removed the dependecy on Date::Utils, no longer required.

1.19  Sat Jun 20 19:40:00 2015
      - Fixed the bug in the method from_gregorian().

1.18  Fri Jun 12 10:00:00 2015
      - Updated method _calendar() to validate month and year.

1.17  Mon May 25 14:40:00 2015
      - Dropped the dependency on Date::Utils::Saka.
      - Upgraded the dependency on Date::Saka::Simple v0.05.
      - Updated code w.r.t the above changes.

1.16  Sat May 02 20:50:00 2015
      - Added LICENSE file.

1.15  Thu Apr 23 23:45:00 2015
      - Re-structured and consumed Date::Utils::Saka as Moo Role.

1.14  Tue Apr 21 13:50:10 2015
      - Fixed the bug in the method _calendar() the last cell is empty.

1.13  Tue Apr 21 10:20:10 2015
      - Complete re-write using Moo with unit tests.

1.12  Fri Jan 16 15:20:10 2015
      - Removed SIGNATURE file.
      - Updated MANIFEST.SKIP file to ignore folder 'eumm/'.
      - Updated Copyright year information.

1.11  Fri Sep 12 22:30:19 2014
      - Deleted Build.PL file.
      - Removed dependency of Readonly.pm and Carp.pm.
      - Added section "REPOSITORY" to the pod
      - Added key "META_MERGE" to Makefile.PL

1.10  Tue Aug 09 16:32:11 2011
      - Made Readonly v1.03 as minimum requirement.
      - Prepared the distribution using the Module::Signature v0.68.

0.09  Tue Apr 19 10:42:55 2011
      - Fixed the typo in the pod.

0.08  Tue Apr 19 09:36:29 2011
      - Renamed tilte "SYNOPSIS" to "DESCRIPTION".

0.07  Mon Apr 18 17:22:11 2011
      - Updated description.
      - Renamed method days_in_year_month() to days_in_month().
      - Updated unit test 01-test-case.t regarding method name change.

0.06  Mon Apr 18 16:54:00 2011
      - Renamed title "Month Names" to "MONTHS" and set it head1.
      - Removed title "Saka Calendar for the month of Phalgun year 1932.".
      - Cosmetics changes.

0.05  Tue Apr 12 12:05:00 2011
      - Fixed buggy internal method _validate_date().
      - Added unit test 05-test-case.t

0.04  Tue Mar 29 10:53:10 2011
      - Added more information about the month name and weekdays of Saka calendar.

0.03  Fri Mar 18 14:15:16 2011
      - Corrected typo in the SYNOPSIS.

0.02  Fri Mar 18 12:29:00 2011
      - Renamed month "Caitra" to "Chaitra".
      - Updated SYNOPSIS to tell more about Saka Calendar.
      - Renamed method _days_in_caitra() to _days_in_chaitra().
      - Updated method to_julian() and from_julian() for month name "Chaitra".
      - Cleaned up method as_string() and method get_calendar().
      - Updated test 04-test-case.t for month name "Chaitra".
      - Updated pod for method days_in_year_month().

0.01  Thu Mar 17 16:10:29 2011
      - First version released.

Makefile.PL
use 5.006;
use strict;
use warnings FATAL => 'all';
use ExtUtils::MakeMaker;

WriteMakefile(
    NAME               => 'Calendar::Saka',
    AUTHOR             => q{Mohammad S Anwar <mohammad.anwar@yahoo.com>},
    VERSION_FROM       => 'lib/Calendar/Saka.pm',
    ABSTRACT_FROM      => 'lib/Calendar/Saka.pm',
    LICENSE            => 'artistic_2',
    MIN_PERL_VERSION   => 5.006,
    CONFIGURE_REQUIRES => {
        'ExtUtils::MakeMaker' => 0,
    },
    BUILD_REQUIRES => {
        'Test::More'        => 0,
        'File::Temp'        => 0,
        'XML::SemanticDiff' => 0,
    },
    PREREQ_PM => {
        'Moo'                        => '2.000000',
        'namespace::clean'           => '0.24',
        'Date::Saka::Simple'         => '0.17',
        'Calendar::Plugin::Renderer' => '0.12',
    },
    dist  => { COMPRESS => 'gzip -9f', SUFFIX => 'gz', },
    clean => { FILES => 'Calendar-Saka-*' },
    (eval { ExtUtils::MakeMaker->VERSION(6.46) } ? (META_MERGE => {
        'meta-spec' => { version => 2 },
        provides    => {
            'Calendar::Saka' => { file => 'lib/Calendar/Saka.pm', version => '1.33' },
        },
        resources => {
            repository => {
                type => 'git',
                url  => 'https://github.com/manwar/Calendar-Saka.git',
                web  => 'https://github.com/manwar/Calendar-Saka',
            },
        }})
     : ()
    ),
);


lib/Calendar/Saka.pm
package Calendar::Saka;

$Calendar::Saka::VERSION   = '1.33';
$Calendar::Saka::AUTHORITY = 'cpan:MANWAR';

=head1 NAME

Calendar::Saka - Interface to Indian Calendar.

=head1 VERSION

Version 1.33

=cut

use 5.006;
use Data::Dumper;

use Date::Saka::Simple;
use Moo;
use namespace::clean;
with 'Calendar::Plugin::Renderer';

use overload q{""} => 'as_string', fallback => 1;

has year  => (is => 'rw', predicate => 1);
has month => (is => 'rw', predicate => 1);
has date  => (is => 'ro', default   => sub { Date::Saka::Simple->new });

sub BUILD {
    my ($self) = @_;

    $self->date->validate_year($self->year)   if $self->has_year;
    $self->date->validate_month($self->month) if $self->has_month;

    unless ($self->has_year && $self->has_month) {
        $self->year($self->date->year);
        $self->month($self->date->month);
    }
}

=head1 DESCRIPTION

Module  to  play  with Saka calendar  mostly  used  in  the South indian, Goa and
Maharashatra. It supports the functionality to add / minus days, months and years
to a Saka date. It can also converts Saka date to Gregorian/Julian date.

The  Saka eras are lunisolar calendars, and feature annual cycles of twelve lunar
months, each month divided into two phases:   the  'bright half' (shukla) and the
'dark half'  (krishna);  these correspond  respectively  to  the  periods  of the
'waxing' and the 'waning' of the moon. Thus, the  period beginning from the first
day  after  the new moon  and  ending on the full moon day constitutes the shukla
paksha or 'bright half' of the month the period beginning from the  day after the
full moon until &  including the next new moon day constitutes the krishna paksha
or 'dark half' of the month.

The  "year zero"  corresponds  to  78 BCE in the Saka calendar. The Saka calendar
begins with the month of Chaitra (March) and the Ugadi/Gudi Padwa festivals  mark
the new year.

Each  month  in  the Shalivahana  calendar  begins with the  'bright half' and is
followed by the 'dark half'.  Thus,  each  month of the Shalivahana calendar ends
with the no-moon day and the new month begins on the day after that.

A variant of the Saka Calendar was reformed & standardized as the Indian National
calendar in 1957. This official  calendar follows the Shalivahan Shak calendar in
beginning from the month of Chaitra and counting years with 78 CE being year zero.
It features a constant number of days in every month with leap years.Saka Calendar
for the month of Chaitra year 1937

    +----------------------------------------------------------------------------------------------------------------------+
    |                                                 Chaitra    [1937 BE]                                                 |
    +----------------+----------------+----------------+----------------+----------------+----------------+----------------+
    |       Ravivara |        Somvara |    Mangalavara |      Budhavara | Brahaspativara |      Sukravara |       Sanivara |
    +----------------+----------------+----------------+----------------+----------------+----------------+----------------+
    |              1 |              2 |              3 |              4 |              5 |              6 |              7 |
    +----------------+----------------+----------------+----------------+----------------+----------------+----------------+
    |              8 |              9 |             10 |             11 |             12 |             13 |             14 |
    +----------------+----------------+----------------+----------------+----------------+----------------+----------------+
    |             15 |             16 |             17 |             18 |             19 |             20 |             21 |
    +----------------+----------------+----------------+----------------+----------------+----------------+----------------+
    |             22 |             23 |             24 |             25 |             26 |             27 |             28 |
    +----------------+----------------+----------------+----------------+----------------+----------------+----------------+
    |             29 |             30 |                                                                                    |
    +----------------+----------------+----------------+----------------+----------------+----------------+----------------+

The package L<App::calendr> provides command line tool  C<calendr> to display the
supported calendars on the terminal.

=head1 SYNOPSIS

    use strict; use warnings;
    use Calendar::Saka;

    # prints current saka month calendar.
    print Calendar::Saka->new->current, "\n";

    # prints saka month calendar in which the given gregorian date falls in.
    print Calendar::Saka->new->from_gregorian(2015, 4, 19), "\n";

    # prints saka month calendar in which the given julian date falls in.
    print Calendar::Saka->new->from_julian(2457102.5), "\n";

    # prints current month saka calendar in SVG format.
    print Calendar::Saka->new->as_svg;

    # prints current month saka calendar in text format.
    print Calendar::Saka->new->as_text;

=head1 SAKA MONTHS

    +-------+-------------------------------------------------------------------+
    | Order | Name                                                              |
    +-------+-------------------------------------------------------------------+
    |   1   | Chaitra                                                           |
    |   2   | Vaisakha                                                          |
    |   3   | Jyaistha                                                          |
    |   4   | Asadha                                                            |
    |   5   | Sravana                                                           |
    |   6   | Bhadra                                                            |
    |   7   | Asvina                                                            |
    |   8   | Kartika                                                           |
    |   9   | Agrahayana                                                        |
    |  10   | Pausa                                                             |
    |  11   | Magha                                                             |
    |  12   | Phalguna                                                          |
    +-------+-------------------------------------------------------------------+

=head1 SAKA DAYS

    +---------+-----------+-----------------------------------------------------+
    | Weekday | Gregorian | Saka                                                |
    +---------+-----------+-----------------------------------------------------+
    |    0    | Sunday    | Ravivara                                            |
    |    1    | Monday    | Somvara                                             |
    |    2    | Tuesday   | Mangalavara                                         |
    |    3    | Wednesday | Budhavara                                           |
    |    4    | Thursday  | Brahaspativara                                      |
    |    5    | Friday    | Sukravara                                           |
    |    6    | Saturday  | Sanivara                                            |
    +---------+-----------+-----------------------------------------------------+

=head1 METHODS

=head2 current()

Returns current month of the Saka calendar.

=cut

sub current {
    my ($self) = @_;

    return $self->as_text($self->date->month, $self->date->year);
}

=head2 from_gregorian()

Returns saka month calendar in which the given gregorian date falls in.

=cut

sub from_gregorian {
    my ($self, $year, $month, $day) = @_;

    return $self->from_julian($self->date->gregorian_to_julian($year, $month, $day));
}

=head2 from_julian($julian_date)

Returns saka month calendar in which the given julian date falls in.

=cut

sub from_julian {
    my ($self, $julian) = @_;

    my $date = $self->from_julian($julian);
    return $self->as_text($date->month, $date->year);
}

=head2 as_svg($month, $year)

Returns calendar for the given C<$month> and C<$year> rendered  in SVG format. If
C<$month> and C<$year> missing, it would return current calendar month.

=cut

sub as_svg {
    my ($self, $month, $year) = @_;

    ($month, $year) = $self->validate_params($month, $year);
    my $date = Date::Saka::Simple->new({ year => $year, month => $month, day => 1 });

    return $self->svg_calendar(
        {
            start_index => $date->day_of_week,
            month_name  => $date->get_month_name,
            days        => $date->days_in_month_year($month, $year),
            year        => $year
        });
}

=head2 as_text($month, $year)

Returns  color  coded  Saka  calendar  for  the  given C<$month> and C<$year>. If
C<$month> and C<$year> missing, it would return current calendar month.

=cut

sub as_text {
    my ($self, $month, $year) = @_;

    ($month, $year) = $self->validate_params($month, $year);
    my $date = Date::Saka::Simple->new({ year => $year, month => $month, day => 1 });

    return $self->text_calendar(
        {
            start_index => $date->day_of_week,
            month_name  => $date->get_month_name,
            days        => $date->days_in_month_year($month, $year),
            day_names   => $date->days,
            year        => $year
        });
}

sub as_string {
    my ($self) = @_;

    return $self->as_text($self->month, $self->year);
}

=head1 AUTHOR

Mohammad S Anwar, C<< <mohammad.anwar at yahoo.com> >>

=head1 REPOSITORY

L<https://github.com/manwar/Calendar-Saka>

=head1 BUGS

Please  report any bugs or feature requests to C<bug-calendar-saka at rt.cpan.org>,
or through the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Calendar-Saka>.
I will be notified, and then you'll automatically be notified of progress on your
bug as I make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Calendar::Saka

You can also look for information at:

=over 4

=item * RT: CPAN's request tracker

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Calendar-Saka>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/Calendar-Saka>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/Calendar-Saka>

=item * Search CPAN

L<http://search.cpan.org/dist/Calendar-Saka/>

=back

=head1 LICENSE AND COPYRIGHT

Copyright (C) 2011 - 2016 Mohammad S Anwar.

This  program  is  free software; you can redistribute it and/or modify it under
the  terms  of the the Artistic License (2.0). You may obtain a copy of the full
license at:

L<http://www.perlfoundation.org/artistic_license_2_0>

Any  use,  modification, and distribution of the Standard or Modified Versions is
governed by this Artistic License.By using, modifying or distributing the Package,
you accept this license. Do not use, modify, or distribute the Package, if you do
not accept this license.

If your Modified Version has been derived from a Modified Version made by someone
other than you,you are nevertheless required to ensure that your Modified Version
 complies with the requirements of this license.

This  license  does  not grant you the right to use any trademark,  service mark,
tradename, or logo of the Copyright Holder.

This license includes the non-exclusive, worldwide, free-of-charge patent license
to make,  have made, use,  offer to sell, sell, import and otherwise transfer the
Package with respect to any patent claims licensable by the Copyright Holder that
are  necessarily  infringed  by  the  Package. If you institute patent litigation
(including  a  cross-claim  or  counterclaim) against any party alleging that the
Package constitutes direct or contributory patent infringement,then this Artistic
License to you shall terminate on the date that such litigation is filed.

Disclaimer  of  Warranty:  THE  PACKAGE  IS  PROVIDED BY THE COPYRIGHT HOLDER AND
CONTRIBUTORS  "AS IS'  AND WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES. THE IMPLIED
WARRANTIES    OF   MERCHANTABILITY,   FITNESS   FOR   A   PARTICULAR  PURPOSE, OR
NON-INFRINGEMENT ARE DISCLAIMED TO THE EXTENT PERMITTED BY YOUR LOCAL LAW. UNLESS
REQUIRED BY LAW, NO COPYRIGHT HOLDER OR CONTRIBUTOR WILL BE LIABLE FOR ANY DIRECT,
INDIRECT, INCIDENTAL,  OR CONSEQUENTIAL DAMAGES ARISING IN ANY WAY OUT OF THE USE
OF THE PACKAGE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

=cut

1; # End of Calendar::Saka
